lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,268 @@
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+
+Options:
+BANNER
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        parsed['certnames'] = []
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--certname FOO,BAR', Array,
+               'One or more comma separated certnames') do |certs|
+            parsed['certnames'] += certs
+          end
+          opts.on('--help', 'Display this create specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--config CONF', 'Path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+        end
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        errors = Utils.parse_with_errors(parser, args)
+
+        if results['certnames'].empty?
+          errors << '    At least one certname is required to create'
+        else
+          results['certnames'].each do |certname|
+            if CERTNAME_BLACKLIST.include?(certname)
+              errors << "    Cannot manage cert named `#{certname}` from " +
+                        "the CLI, if needed use the HTTP API directly"
+            end
+
+            unless certname =~ VALID_CERTNAME
+              errors << "  Certname #{certname} must not contain unprintable or non-ASCII characters"
+            end
+          end
+        end
+
+        errors_were_handled = Utils.handle_errors(@logger, errors, parser.help)
+
+        exit_code = errors_were_handled ? 1 : nil
+
+        return results, exit_code
+      end
+
+      def run(input)
+        certnames = input['certnames']
+        config_path = input['config']
+
+        # Validate config_path provided
+        if config_path
+          errors = FileUtilities.validate_file_paths(config_path)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        # Load most secure signing digest we can for csr signing.
+        signer = SigningDigest.new
+        return 1 if Utils.handle_errors(@logger, signer.errors)
+
+        # Make sure we have all the directories where we will be writing files
+        FileUtilities.ensure_dir(puppet.settings[:certdir])
+        FileUtilities.ensure_dir(puppet.settings[:privatekeydir])
+        FileUtilities.ensure_dir(puppet.settings[:publickeydir])
+
+        # Generate and save certs and associated keys
+        all_passed = generate_certs(certnames, puppet.settings, signer.digest)
+        return all_passed ? 0 : 1
+      end
+
+      # Create crls and keys, then submit them to CA, request for the CA to sign
```

Comment:
Create _certs_ and keys   ... not _crls_

Reply:
thanks.  will fix

Reply:
^ Should have been csrs